ccl/changefeedccl/schemafeed: TestTableHistoryIngestionTracking failed

ccl/changefeedccl/schemafeed.TestTableHistoryIngestionTracking [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsGcpLinuxX8664BigVm_CclUnitTests/13075157?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsGcpLinuxX8664BigVm_CclUnitTests/13075157?buildTab=artifacts#/) on release-23.2 @ [43a2532550ca7abfc3c5682df31b0cc20da934b0](https://github.com/cockroachdb/cockroach/commits/43a2532550ca7abfc3c5682df31b0cc20da934b0):

Fatal error:

```
panic: test timed out after 4m55s
running tests:
	TestTableHistoryIngestionTracking (4m55s)

```
Stack: 

```
goroutine 31 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2241 +0x3c5
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x32
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTableHistoryIngestionTracking
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTableHistoryIngestionTracking.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-34377

--- a/pkg/ccl/changefeedccl/schemafeed/schema_feed.go
+++ b/pkg/ccl/changefeedccl/schemafeed/schema_feed.go
@@ -523,29 +523,16 @@ func (tf *schemaFeed) highWater() hlc.Timestamp {
 // `validateFn` is deterministic and the ingested descriptors are read
 // transactionally).
 func (tf *schemaFeed) waitForTS(ctx context.Context, ts hlc.Timestamp) error {
-	var errCh chan error
+	waitCh, feedErr := tf.tryWaitForTS(ts)
+	if feedErr != nil {
+		return feedErr
+	}
 
-	highWater := tf.highWater()
-	var err error
-	var fastPath bool
-	func() {
-		tf.mu.Lock()
-		defer tf.mu.Unlock()
-		if !tf.mu.errTS.IsEmpty() && tf.mu.errTS.LessEq(ts) {
-			err = tf.mu.err
-		}
-		fastPath = err != nil || ts.LessEq(highWater)
-		if !fastPath {
-			// non-fastPath is when we need to prove the invariant holds from [`high_water`, `ts].
-			errCh = make(chan error, 1)
-			tf.mu.waiters = append(tf.mu.waiters, tableHistoryWaiter{ts: ts, errCh: errCh})
-		}
-	}()
-	if fastPath {
+	if waitCh == nil {
 		if log.V(1) {
-			log.Infof(ctx, "fastpath for %s: %v", ts, err)
+			log.Infof(ctx, "fastpath for %s", ts)
 		}
-		return err
+		return nil
 	}
 
 	if log.V(1) {
@@ -555,7 +542,7 @@ func (tf *schemaFeed) waitForTS(ctx context.Context, ts hlc.Timestamp) error {
 	select {
 	case <-ctx.Done():
 		return ctx.Err()
-	case err := <-errCh:
+	case err := <-waitCh:
 		if log.V(1) {
 			log.Infof(ctx, "waited %s for %s highwater: err=%v", timeutil.Since(start), ts, err)
 		}
@@ -564,6 +551,27 @@ func (tf *schemaFeed) waitForTS(ctx context.Context, ts hlc.Timestamp) error {
 		}
 		return err
 	}
+}
+
+// tryWaitForTS is a fast path for waitForTS.  Returns non-nil channel
+// if the fast path not available.
+func (tf *schemaFeed) tryWaitForTS(ts hlc.Timestamp) (chan error, error) {
+	tf.mu.Lock()
+	defer tf.mu.Unlock()
+
+	if !tf.mu.errTS.IsEmpty() && tf.mu.errTS.LessEq(ts) {
+		// Schema feed error occurred.
+		return nil, tf.mu.err
+	}
+
+	if ts.LessEq(tf.mu.highWater) {
+		return nil, nil // Fast path.
+	}
+
+	// non-fastPath is when we need to prove the invariant holds from [`high_water`, `ts].
+	waitCh := make(chan error, 1)
+	tf.mu.waiters = append(tf.mu.waiters, tableHistoryWaiter{ts: ts, errCh: waitCh})
+	return waitCh, nil
 }
 
 // descLess orders descriptors by (modificationTime, id).

